File: backend/models/meal.py
"""Meal (feeding) model + DTOs.

Each meal row is an immutable snapshot of *what was eaten*. At log time we
copy the food's per-100g macros, multiply by `actual_weight_g`, and store the
absolute gram values on the meal row. Editing the source food later does NOT
retroactively change historical meals.

`serving_type` lets the client send either a count of natural units (cans,
cups) or a direct gram weight; `actual_weight_g` is what we calculate and
all macros are derived from it.
"""

from datetime import datetime
from enum import Enum

from pydantic import BaseModel, Field


meal_table = "meals"


class MealType(str, Enum):
    BREAKFAST = "breakfast"
    LUNCH = "lunch"
    DINNER = "dinner"
    SNACK = "snack"


class ServingType(str, Enum):
    UNITS = "units"  # natural units (cans, cups) — multiplied by food.unit_weight
    GRAMS = "grams"  # direct gram weight


# ────── Stored row shape ──────


class Meal(BaseModel):
    id: str
    pet_id: str
    food_id: str
    user_id: str
    group_id: str
    timestamp: datetime
    meal_type: MealType | None = None
    serving_type: ServingType
    serving_amount: float
    actual_weight_g: float
    calories: float
    protein_g: float
    fat_g: float
    moisture_g: float
    carbohydrate_g: float
    notes: str | None = None
    is_active: bool = True
    created_at: datetime
    updated_at: datetime


# ────── Request DTOs ──────


class CreateMealRequest(BaseModel):
    pet_id: str
    food_id: str
    serving_type: ServingType
    serving_amount: float = Field(..., gt=0, le=10000)
    timestamp: datetime | None = None  # defaults to "now" in service
    meal_type: MealType | None = None
    notes: str | None = Field(None, max_length=500)


class UpdateMealRequest(BaseModel):
    meal_id: str
    food_id: str | None = None
    timestamp: datetime | None = None
    meal_type: MealType | None = None
    serving_type: ServingType | None = None
    serving_amount: float | None = Field(None, gt=0, le=10000)
    notes: str | None = Field(None, max_length=500)


class DeleteMealRequest(BaseModel):
    meal_id: str


# ────── Response DTOs ──────


class MealSummary(BaseModel):
    """Lightweight row for list endpoints."""

    id: str
    pet_id: str
    pet_name: str
    food_id: str
    food_brand: str
    food_product_name: str
    user_id: str
    fed_by_name: str
    group_id: str
    timestamp: datetime
    meal_type: MealType | None = None
    serving_type: ServingType
    serving_amount: float
    actual_weight_g: float
    calories: float
    created_at: datetime
    updated_at: datetime


class MealDetails(BaseModel):
    id: str
    pet_id: str
    pet_name: str
    food_id: str
    food_brand: str
    food_product_name: str
    user_id: str
    fed_by_name: str
    group_id: str
    group_name: str
    timestamp: datetime
    meal_type: MealType | None = None
    serving_type: ServingType
    serving_amount: float
    actual_weight_g: float
    calories: float
    protein_g: float
    fat_g: float
    moisture_g: float
    carbohydrate_g: float
    notes: str | None = None
    is_active: bool
    created_at: datetime
    updated_at: datetime


class TodayMealsResponse(BaseModel):
    """Today's totals + the meal rows themselves so the client can render in one shot."""

    date: str  # YYYY-MM-DD (the day the totals are scoped to)
    total_meals: int
    total_calories: float
    total_weight_g: float
    breakfast_count: int
    lunch_count: int
    dinner_count: int
    snack_count: int
    pet_id: str | None = None
    pet_name: str | None = None
    daily_calorie_target: int | None = None
    calorie_target_percentage: float | None = None
    group_id: str | None = None
    pets_fed_count: int | None = None
    meals: list[MealSummary]


class MealStatistics(BaseModel):
    date_from: str
    date_to: str
    total_days: int
    total_meals: int
    total_calories: float
    total_weight_g: float
    average_meals_per_day: float
    average_calories_per_day: float
    average_protein_g_per_day: float
    average_fat_g_per_day: float
    average_moisture_g_per_day: float
    average_carbohydrate_g_per_day: float
    meal_type_counts: dict[str, int]

File: frontend/src/types/meal.ts
// Meal Types

export type MealType = 'breakfast' | 'lunch' | 'dinner' | 'snack';
export type ServingType = 'units' | 'grams';

export interface MealRecord {
  id: string;
  pet_id: string;
  food_id: string;
  user_id: string;
  group_id: string;
  timestamp: string; // ISO date string
  meal_type: MealType | null;
  serving_type: ServingType;
  serving_amount: number;
  actual_weight_g: number;
  calories: number;
  protein_g: number;
  fat_g: number;
  moisture_g: number;
  carbohydrate_g: number;
  notes?: string;
  created_at: string;
  updated_at: string;
  is_active: boolean;
}

export interface MealInfo {
  id: string;
  pet_id: string;
  pet_name: string;
  food_id: string;
  food_name: string;
  user_id: string;
  fed_by_name: string;
  timestamp: string;
  meal_type: MealType | null;
  serving_type: ServingType;
  serving_amount: number;
  actual_weight_g: number;
  calories: number;
  created_at: string;
  updated_at: string;
  group_id: string;
}

export interface MealDetails {
  id: string;
  pet_id: string;
  pet_name: string;
  food_id: string;
  food_name: string;
  food_brand: string;
  food_product_name: string;
  user_id: string;
  fed_by_name: string;
  group_id: string;
  group_name: string;
  timestamp: string;
  meal_type: MealType | null;
  serving_type: ServingType;
  serving_amount: number;
  actual_weight_g: number;
  calories: number;
  protein_g: number;
  fat_g: number;
  moisture_g: number;
  carbohydrate_g: number;
  created_at: string;
  updated_at: string;
  notes?: string;
}

export interface CreateMealRequest {
  pet_id: string;
  food_id: string;
  fed_at?: string; // ISO date string, defaults to now
  meal_type?: MealType;
  serving_type: ServingType;
  serving_amount: number;
  notes?: string;
}

export interface UpdateMealRequest {
  food_id?: string;
  fed_at?: string;
  meal_type?: MealType;
  serving_type?: ServingType;
  serving_amount?: number;
  notes?: string;
}

export interface TodayMealSummary {
  date: string; // YYYY-MM-DD format
  total_meals: number;
  total_calories: number;
  total_weight_g: number;
  breakfast_count: number;
  lunch_count: number;
  dinner_count: number;
  snack_count: number;
  pet_id?: string;
  pet_name?: string;
  daily_calorie_target?: number;
  calorie_target_percentage?: number;
  group_id?: string;
  pets_fed_count?: number;
}

export interface MealStatistics {
  date_from: string;
  date_to: string;
  total_days: number;
  total_meals: number;
  total_calories: number;
  total_weight_g: number;
  average_meals_per_day: number;
  average_calories_per_day: number;
  average_protein_g_per_day: number;
  average_fat_g_per_day: number;
  average_moisture_g_per_day: number;
  average_carbohydrate_g_per_day: number;
  meal_type_distribution: Record<string, number>;
  most_active_feeders: Array<{ user_name: string; meal_count: number }>;
  most_used_foods: Array<{ food_name: string; usage_count: number }>;
  target_achievement?: {
    days_on_target: number;
    average_target_percentage: number;
  };
}

export interface MealQueryFilters {
  pet_id?: string;
  group_id?: string;
  user_id?: string;
  date_from?: string; // YYYY-MM-DD format
  date_to?: string; // YYYY-MM-DD format
  meal_type?: MealType;
  limit?: number;
  offset?: number;
}
